File: packages/js/src/plugins/haptics.ts
import { NativePlugin, Plugin } from '../plugin';

export interface VibrateOptions {
  duration?: number;
}

@NativePlugin({
  name: 'Haptics',
  id: 'com.avocadojs.plugin.haptics'
})
export class Haptics extends Plugin {

  impact(options: HapticsImpactOptions) {
    this.nativeCallback('impact', options);
  }

  vibrate(options: VibrateOptions = {}) {
    this.nativeCallback('vibrate', options);
  }

  selectionStart() {
    this.nativeCallback('selectionStart');
  }

  selectionChanged() {
    this.nativeCallback('selectionChanged');
  }

  selectionEnd() {
    this.nativeCallback('selectionEnd');
  }
}


export interface HapticsImpactOptions {
  style: HapticsImpactStyle;
}

export enum HapticsImpactStyle {
  Heavy = 'HEAVY',
  Medium = 'MEDIUM',
  Light = 'LIGHT'
}

File: ios/Avocado/Avocado/Plugins/Haptics.swift
import Foundation
import AudioToolbox

@objc(Haptics)
public class Haptics: AVCPlugin {
  var selectionFeedbackGenerator: UISelectionFeedbackGenerator?
  
  @objc public func impact(_ call: AVCPluginCall) {
    if let style = call.options["style"] as? String {
      var impactStyle = UIImpactFeedbackStyle.heavy
      if style == "MEDIUM" {
        impactStyle = UIImpactFeedbackStyle.medium
      } else if style == "LIGHT" {
        impactStyle = UIImpactFeedbackStyle.light
      }
      let generator = UIImpactFeedbackGenerator(style: impactStyle)
      generator.impactOccurred()
    } else {
      let generator = UIImpactFeedbackGenerator(style: .heavy)
      generator.impactOccurred()
    }
  }
  
  @objc public func selectionStart(_ call: AVCPluginCall) {
    selectionFeedbackGenerator = UISelectionFeedbackGenerator()
    selectionFeedbackGenerator?.prepare()
  }
  
  @objc public func selectionChanged(_ call: AVCPluginCall) {
    if let generator = selectionFeedbackGenerator {
      generator.selectionChanged()
      generator.prepare()
    }
  }
  
  @objc public func selectionEnd(_ call: AVCPluginCall) {
    selectionFeedbackGenerator = nil
  }
  
  @objc public func vibrate(_ call: AVCPluginCall) {
    AudioServicesPlayAlertSound(kSystemSoundID_Vibrate)
  }
}
